zipline/pipeline/domain.py

```diff
@@ -0,0 +1,182 @@
+"""
+Domains
+-------
+
+TODO_SS
+"""
+from interface import implements, Interface
+
+from trading_calendars import get_calendar
+
+from zipline.country import CountryCode
+from zipline.utils.memoize import lazyval
+
+from .sentinels import NotSpecified
+
+
+class IDomain(Interface):
```

Comment:
I think this is useful for keeping the testing implementations in sync with the main implementation. I've already gone through at least one major revision of what methods I needed here. Initially these provided a calendar instead of just sessions, but that made it hard to port some of the blaze core loader tests (see note on https://github.com/quantopian/zipline/pull/2266/files/0a6405a1b65461752e10d8651447b070879a1e66#diff-2e4c20f8c11194598b15ac870e83ff15R31), so I changed to have this just provide a session index, which is (at least right now) all we need.

The one thing I'm not super excited about on this is that we actually care about types here because we want to be able to validate user input, and at least on quantopian I don't think we want to support duck-typed domains. But I think even with the slightly awkward workaround the extra strictness of interface makes it a better `abc` than `abc` for what we want here.